Pattern parsing for lists

The following should be a valid pattern
```
| (head::firsttail)::finaltail ->
```
The pattern has type `[[a]]` for some type `a`, where `head` is the head of the first list in the overall list, `firsttail` is the tail of the first list in the overall list, and `finaltail` is the tail of the overall list.  However, it gives the following syntax error (this is at the beginning of line 47):
```
ProofState.sv:47:18: error: Syntax error:
Error at line 47, column 18 in file ProofState.sv
         (parser state: 707; real character index: 1859):
  Expected a token of one of the following types:
   [silver:compiler:definition:core:BlockComments,
    '::' (silver:compiler:definition:core:ColonColon_t),
    ',' (silver:compiler:definition:core:Comma_t),
    silver:compiler:definition:core:Comments,
    silver:compiler:definition:core:WhiteSpace]
   Input currently matches:
   [silver:compiler:extension:templating:syntax:QuoteWater,
    silver:compiler:extension:templating:syntax:SingleLineQuoteWater]
```

This does not seem to be a problem for expressions, since the expression
```
(head::firsttail)::finaltail
```
parses fine as an expression instead of as a pattern.

--- a/grammars/silver/compiler/extension/patternmatching/PatternTypes.sv
+++ b/grammars/silver/compiler/extension/patternmatching/PatternTypes.sv
@@ -200,6 +200,7 @@ top::Pattern ::=
 concrete production nestedPatterns
 top::Pattern ::= '(' p::Pattern ')'
 {
+  top.unparse = s"(${p.unparse})";
   forwards to p;
 }
 

--- a/grammars/silver/compiler/extension/patternmatching/PatternTypes.sv
+++ b/grammars/silver/compiler/extension/patternmatching/PatternTypes.sv
@@ -196,6 +196,13 @@ top::Pattern ::=
   top.patternTypeName = "";
 }
 
+-- arbitrary nesting of patterns
+concrete production nestedPatterns
+top::Pattern ::= '(' p::Pattern ')'
+{
+  forwards to p;
+}
+
 --------------------------------------------------------------------------------
 
 -- Below are the non-canonical patterns, i.e. those for other types
@@ -390,8 +397,6 @@ top::NamedPattern ::= qn::QName '=' p::Pattern
   top.patternVars = p.patternVars;
   top.namedPatternList = [pair(qn.lookupAttribute.fullName, p)];
 }
-
-
 
 --helper function for building patternLists from lists of patterns
 function buildPatternList

--- a/test/patt/Nested.sv
+++ b/test/patt/Nested.sv
@@ -60,3 +60,17 @@ equalityTest ( nestedprods( nest_empty() ), "nest_empty", String, pat_tests ) ;
 equalityTest ( nestedprods( nest_int(2) ), "nest_int_2", String, pat_tests ) ;
 equalityTest ( nestedprods( nest_double( nest_empty(), nest_int(2)) ), "nest_double_specific", String, pat_tests ) ;
 equalityTest ( nestedprods( nest_double( nest_int(3), nest_empty()) ), "nest_double_general", String, pat_tests ) ;
+
+
+wrongCode "1.0 is a Float but we're trying to match against Integer" {
+ nonterminal CompletelyValidNonterminal;
+ 
+ function funfoo
+ String ::= n::Integer
+ {
+  return case n of
+         | (1.0) -> "Fail"
+         | _ -> "OK"
+         end;
+ }
+}

--- a/test/patt/Nested.sv
+++ b/test/patt/Nested.sv
@@ -1,0 +1,62 @@
+grammar patt;
+
+-- Fixing #495 to support arbitrary nesting of patterns
+function nestedlists
+String ::= alist :: [[Integer]]
+{
+ return case alist of
+        | (head::firsttail)::finaltail -> "NotEmpty/NotEmpty"
+        | _::finaltail -> "Empty/NotEmpty"
+        | [] -> "Empty" 
+        end;
+}
+
+function nestedints
+Integer ::= i ::Integer
+{
+ return case i of
+        | 0 -> 10
+        | (1) -> 11
+        | ((2)) -> 12
+        | _ -> 0
+        end;
+}
+
+nonterminal Nest;
+production nest_empty
+n::Nest ::= 
+{ }
+
+production nest_int
+n::Nest ::= i::Integer
+{ }
+
+production nest_double
+n::Nest ::= n1::Nest n2::Nest
+{ }
+
+function nestedprods
+String ::= n::Nest
+{
+  return case n of
+         | (nest_empty()) -> "nest_empty"
+         | ((nest_int((2)))) -> "nest_int_2"
+         | (nest_int((_))) -> "nest_int"
+         | ((nest_double((nest_empty()),(nest_int((_)))))) -> "nest_double_specific"
+         | (((nest_double(((_)),_)))) -> "nest_double_general"
+         end;
+}
+
+equalityTest ( nestedlists( [ [1,2] , [3,4] ] ), "NotEmpty/NotEmpty", String, pat_tests ) ;
+equalityTest ( nestedlists( [ [] , [3,4] ] ), "Empty/NotEmpty", String, pat_tests ) ;
+equalityTest ( nestedlists( [ ] ), "Empty", String, pat_tests ) ;
+
+equalityTest ( nestedints( 0 ), 10, Integer, pat_tests ) ;
+equalityTest ( nestedints( 1 ), 11, Integer, pat_tests ) ;
+equalityTest ( nestedints( 2 ), 12, Integer, pat_tests ) ;
+equalityTest ( nestedints( 3 ),  0, Integer, pat_tests ) ;
+
+equalityTest ( nestedprods( nest_empty() ), "nest_empty", String, pat_tests ) ;
+equalityTest ( nestedprods( nest_int(2) ), "nest_int_2", String, pat_tests ) ;
+equalityTest ( nestedprods( nest_double( nest_empty(), nest_int(2)) ), "nest_double_specific", String, pat_tests ) ;
+equalityTest ( nestedprods( nest_double( nest_int(3), nest_empty()) ), "nest_double_general", String, pat_tests ) ;

